Your Giving Component

Omicron7 opened the pull request.

Omicron7 commented:
This is far enough along I'd like to merge it in.
Donor overview and recipient views are working. Still need to implement the Monthly view.

OzzieOrca commented on src/app/profile/yourGiving/givingRecipientView/donationRecipient/donationRecipient.tpl.html:
```diff
@@ -0,0 +1,88 @@
+            <div class="form-group mb0" ng-if="$ctrl.recipient['mostrecentdonationRecurringdonationelement']">
+              <label translate>Recurring Gifts:</label>
+              <ul class="list-unstyled list-recurring">
+                <li ng-repeat="recurring in $ctrl.recipient['mostrecentdonationRecurringdonationelement']">
```
My account has data from the old prod site that must have been imported a year or two ago.  For some reason I have lots of these recurring gifts of the same number and date. Idk if that is an issue with it being old imported data or the new backend implementation.

![screen shot 2016-10-03 at 10 56 10 am](https://cloud.githubusercontent.com/assets/756501/19047763/0a41480c-8958-11e6-8c39-011f69468b71.png)

```json
"_recurringdonationelement": [
    {
        "self": {
            "type": "donations.recurring",
            "uri": "/donations/recurring/crugive/gewuuwktkbefq=",
            "href": "https://cortex-gateway-stage.cru.org/cortex/donations/recurring/crugive/gewuuwktkbefq="
        },
        "links": [
            {
                "rel": "paymentmethods",
                "uri": "/paymentmethods/crugive",
                "href": "https://cortex-gateway-stage.cru.org/cortex/paymentmethods/crugive"
            },
            {
                "rel": "paymentmethod",
                "uri": "/paymentmethods/crugive/m5uxszddnv5hsz3fhu=",
                "href": "https://cortex-gateway-stage.cru.org/cortex/paymentmethods/crugive/m5uxszddnv5hsz3fhu="
            }
        ],
        "donation-lines": [
            {
                "amount": 15,
                "designation-name": "Test Person (1234567)",
                "designation-number": "1234567",
                "donation-line-row-id": "1-JYSPHZ",
                "donation-line-status": "Standard",
                "payment-method-id": "giydcmzygf=",
                "updated-donation-line-status": "",
                "updated-payment-method-id": "",
                "updated-rate": {
                    "recurrence": {
                        "interval": ""
                    }
                },
                "updated-recurring-day-of-month": "",
                "updated-start-month": "",
                "updated-start-year": ""
            }
        ],
        "donation-row-id": "1-JYSPHX",
        "donation-status": "Active",
        "effective-status": "Active",
        "next-draw-date": {
            "display-value": "2016-10-20",
            "value": 1476921600000
        },
        "rate": {
            "recurrence": {
                "interval": "Monthly"
            }
        },
        "recurring-day-of-month": "20",
        "start-date": {
            "display-value": "2016-09-30",
            "value": 1475193600000
        }
    },
    {
        "self": {
            "type": "donations.recurring",
            "uri": "/donations/recurring/crugive/gewuuwsjgnkue=",
            "href": "https://cortex-gateway-stage.cru.org/cortex/donations/recurring/crugive/gewuuwsjgnkue="
        },
        "links": [
            {
                "rel": "paymentmethods",
                "uri": "/paymentmethods/crugive",
                "href": "https://cortex-gateway-stage.cru.org/cortex/paymentmethods/crugive"
            },
            {
                "rel": "paymentmethod",
                "uri": "/paymentmethods/crugive/m5uxszddnv5hsz3fhu=",
                "href": "https://cortex-gateway-stage.cru.org/cortex/paymentmethods/crugive/m5uxszddnv5hsz3fhu="
            }
        ],
        "donation-lines": [
            {
                "amount": 15,
                "designation-name": "Test Person (1234567)",
                "designation-number": "1234567",
                "donation-line-row-id": "1-JZI3UD",
                "donation-line-status": "Standard",
                "payment-method-id": "giydcmzyge=",
                "updated-donation-line-status": "",
                "updated-payment-method-id": "",
                "updated-rate": {
                    "recurrence": {
                        "interval": ""
                    }
                },
                "updated-recurring-day-of-month": "",
                "updated-start-month": "",
                "updated-start-year": ""
            }
        ],
        "donation-row-id": "1-JZI3UB",
        "donation-status": "Active",
        "effective-status": "Active",
        "next-draw-date": {
            "display-value": "2016-10-20",
            "value": 1476921600000
        },
        "rate": {
            "recurrence": {
                "interval": "Monthly"
            }
        },
        "recurring-day-of-month": "20",
        "start-date": {
            "display-value": "2016-09-30",
            "value": 1475193600000
        }
    },
    ...
]
```

Omicron7 replied:
I haven't seen anything like that. Data I could find only had one item in "donation-lines", so I may not be displaying it properly, or it wasn't what I thought it was.

OzzieOrca commented on src/app/profile/yourGiving/givingRecipientView/givingRecipientView.tpl.html:
```diff
@@ -0,0 +1,2 @@
+<donation-recipient ng-repeat="recipient in $ctrl.recipients | orderBy:'designation-name'"
```
There are lots of people with a year to date amount of 0 and no other useful data other than name and designation number. Should this only be showing active/recent donations? This could also be an issue with my old data.

Omicron7 replied:
I'm betting this is old data as well as "recent" being defined in the API as last 2 years.

OzzieOrca replied:
K

OzzieOrca commented on src/app/profile/yourGiving/givingRecipientView/donationRecipient/donationRecipient.tpl.html:
```diff
@@ -0,0 +1,88 @@
+        <img class="full-width" desig-src="{{$ctrl.recipient['designation-number']}}">
+      </div>
+      <div class="col-sm-12 col-md-7 col-full-height">
+        <h4 class="mb0">{{$ctrl.recipient['designation-name']}}</h4>
```
The designation number seems to be repeated in the name field. Seems to be an api issue. Could be my old data...

Omicron7 replied:
This is how the API generates the display name.

OzzieOrca replied:
Fun. I think we should create a Jira ticket for it

Omicron7 replied:
I was going to let QA drive that.

OzzieOrca commented on src/app/profile/yourGiving/yourGiving.component.js:
```diff
@@ -0,0 +1,121 @@
+    this.dateFilter = $filter( 'date' );
+  }
+
+  $onInit() {
```
This seems like a bit of code for `$onInit`... Do you think it would make sense to move it to its own method so it can be tested separately? I guess it is mostly initialization. Maybe you could make `initSessionEnforcer` and `initFilters` methods. Idk. Whatever you think.

OzzieOrca commented on src/app/profile/yourGiving/yourGiving.tpl.html:
```diff
@@ -0,0 +1,140 @@
+                  </button>
+                </div>
+
+                <div class="btn-group" uib-dropdown role="group" ng-if="$ctrl.view == 'recipient'">
```
It's a little hard to tell from a user's standpoint whether this is a set of separate pages or a filter. The UX mockup has a button right aligned. Idk if this is that functionality. If it is a filter I think it could use the word Filter or Show or something next to it.

Omicron7 replied:
UX is incorrect. Per Jason Chandler, this is the correct behavior. https://jira.cru.org/browse/EP-1202

Omicron7 commented:
I'm working on Monthly view as well as renaming a bunch of the components. I'll have a push in a bit with these changes addressed.

OzzieOrca commented on src/app/profile/yourGiving/givingRecipientView/givingRecipientView.component.js:
```diff
@@ -0,0 +1,49 @@
+    this.setLoading( {loading: true} );
+    this.recipients = [];
+    if ( angular.isDefined( this.subscriber ) ) this.subscriber.unsubscribe();
+    this.subscriber = this.donationsService.getRecipients( year ).subscribe( ( recipients ) => {
```
I meant to ask, what you are using `this.subscriber` for? Somehow it looks like I didn't save my comment...

Omicron7 replied:
Canceling the current subscriber if filter changes happen while it's waiting on a response. I know this doesn't cancel the http request, but it does ignore the response when it come back. This prevents a race condition where we could end up showing results for the previous filter and no the current one.

OzzieOrca replied:
Oh I missed line 24. Makes sense now.

Omicron7 commented:
I've added the Monthly view. Had to guess on it some since UX hasn't provided the template yet.
Fixed product Config modal saying amount was missing.
Renamed a bunch of the components to make more sense.
